GVFS/GVFS.Common/Maintenance/PrefetchStep.cs

```diff
@@ -291,6 +297,58 @@ private bool TrySchedulePostFetchJob(List<string> packIndexes)
             return false;
         }
 
+        /// <summary>
+        /// Ensure the prefetch pack with most-recent timestamp has an associated
+        /// ".keep" file. This prevents any Git command from deleting the pack.
+        ///
+        /// Delete the previous ".keep" file(s) so that pack can be deleted when they
+        /// are not the most-recent pack.
+        /// </summary>
+        private void UpdateKeepPacks()
+        {
+            if (!this.TryGetMaxGoodPrefetchTimestamp(out long maxGoodTimeStamp, out string error))
+            {
+                return;
+            }
+
+            string prefix = $"prefetch-{maxGoodTimeStamp}-";
+
+            DirectoryItemInfo info = this.Context
+                                         .FileSystem
+                                         .ItemsInDirectory(this.Context.Enlistment.GitPackRoot)
+                                         .Where(item => item.Name.StartsWith(prefix)
+                                                        && string.Equals(Path.GetExtension(item.Name), ".pack", StringComparison.OrdinalIgnoreCase))
+                                         .FirstOrDefault();
+            if (info == null)
+            {
+                this.Context.Tracer.RelatedWarning(this.CreateEventMetadata(), $"Could not find latest prefetch pack, starting with {prefix}");
+                return;
+            }
+
+            string newKeepFile = Path.ChangeExtension(info.FullName, ".keep");
+
+            try
+            {
+                this.Context.FileSystem.WriteAllText(newKeepFile, string.Empty);
+
+                foreach (string keepFile in this.Context
+                                         .FileSystem
+                                         .ItemsInDirectory(this.Context.Enlistment.GitPackRoot)
+                                         .Where(item => item.Name.EndsWith(".keep"))
+                                         .Select(item => item.FullName))
+                {
+                    if (!keepFile.Equals(newKeepFile))
+                    {
+                        this.Context.FileSystem.TryDeleteFile(keepFile);
+                    }
+                }
+            }
+            catch (IOException e)

```

Comment:
I noticed that `WriteAllFile` can also throw `UnauthorizedAccessException` and `SecurityException`, and there are other spots we're catching those exception types in the code.

It seems unlikely we'd hit them here (based on their description), but wanted to see if you thought it'd be worth catching them here as well.

Reply:
*Edit:* I meant to say `WriteAllText` above

Reply:
The general approach I've seen taken in the past is:

- Catch exceptions that we anticipate could happen at runtime that are not indicative of a code bug 
- Log exceptions and exit when there's:
    - An otherwise unhandled exception 
    - An exception that indicates that we probably have a bug in the code.  For example, `ArgumentNullException` or `NotSupportedException` could be prevented with code changes, but `SecurityException` probably could not.

I realize this is a bit fuzzy, as some exceptions (e.g. `DirectoryNotFoundException` might not fall cleaning into one of these categories).

I mentioned these three specifically because we've hit them at various times in the past and I'd imagine that if we were to hit them here it'd be better to fail the job than bring down the mount process.

> If we want to catch those three exactly, then I can instead go and build `TryWriteAllText()` to include that logic instead.

I think either approach (`TryWriteAllText()` or catching the exceptions) is good, so I'd say go with whichever you prefer.